doc/users-guide/users-guide-ipsec.adoc

```diff
@@ -0,0 +1,410 @@
+	        ...process IPsec status event
+	        break;
+
+	}
+}
+-----
+
+===== IPsec Inline Processing
+For SAs operating in `ODP_IPSEC_OP_MODE_INLINE`, inbound processing is
+implicit. The application never sees these packets until after IPsec has
+already decrypted them. As shown in the code sketch above, such packets
+appear as events of subtype `ODP_EVENT_PACKET_IPSEC` and the `flag` field in
+the associated `odp_ipsec_packet_result_t` indicates `inline_mode`.
+
+For outbound IPsec processing, the `odp_ipsec_out_inline()` API operates as
+a "fire and forget" API. A success return code from this call indicates that
+the packet will be encrypted and transmitted to the `odp_pktio_t` indicated
+in the `odp_ipsec_out_inline_param_t` specified at the time of the call without
+any further application involvement. Only if a problem arises will the packet
+be returned to the application with an `odp_ipsec_packet_result_t` indicating
+the nature of the problem.
+
+==== SA Lifetimes
+A fundamental principle of good security is that the keying material
+associated with sessions has a limited lifetime. In effect, keys grow "stale"
+over time or due to being used to encrypt too much data. The metrics used
+to limit effective SA lifetimes are:
+
+* Duration (time)
+* Usage (volume of traffic using the keys)
+
+Associated with each of these metrics are "soft" and "hard" limits. When a
+hard limit is reached, the SA is expired and cannot be used further. To support
+graceful transition to a replacement SA, soft limits are used. A soft limit is
+similar to a "low fuel" warning light on a car. It alerts the application that
+the SA is nearing the end of its useful life and should be renegotiated even
+as the SA continues to work normally.
+
+ODP support for SA limits is based on packet/byte counts. Applications that
+wish to use time-based SA limits may do so on their own using the timing
+facilities that ODP provides. However, since especially with inline IPsec
+processing, the application may not have explicit knowledge of the traffic
+volumes associated with a given SA, support for usage-based limits is
+integrated into ODP IPsec support.
+
+At `odp_ipsec_sa_create()` time, one of the fields in the
+`odp_ipsec_sa_param_t` struct is the `odp_ipsec_lifetime_t` sub-structure.
+This struct allows hard and/or soft limits to be specified in terms of total
+bytes encrypted/decrypted, total packet count, or both. A limit specification
+of 0 indicates no limit for that metric. If either is specified, the limit
+is triggered on whichever occurs first. Given the defined behavior of hard vs.
+soft limits, the soft limits, if used, should always be specified as lower
+than the hard limits. These should be sufficiently lower to enable adequate
+time to switch over to a replacement SA before the hard limit is reached.
+
+As noted, when an SA hard limit is reached the SA immediately enters the
+disabled state and attempts to use it further are failed with an
```

Comment:
Ahem. It does not enter disabled state per se:
- It is an undefined behaviour (iow, an application error) to submit packets to disabled SA
- It is a perfectly valid to submit packets to SA after hard limit overrun (e.g. because other packets might be already queued at this moment).

Reply:
If you let yourself run out of gas, the car can stop at inconvenient times, which is why one pays attention to that low fuel warning light. A hard limit is a hard limit. That's what makes it hard. Any other definition seems confusingly fuzzy and unnecessary.

Reply:
Yep. It is just not a 'disabled' state, because we have separate definition for 'disabled SA'.

Reply:
OK, in v4 I've added a new terminal state `SA_Expired` to the FSM and have updated the doc to say "expired" rather than "disabled". From the expired state the only valid operation is `odp_ipsec_sa_destroy()`.